feat(unstable): `deno run --env`

iuioiua opened the pull request:
This change adds the `--env=[FILE]` flag to the `run`, `compile`, `eval`, `install` and `repl` subcommands. Environment variables set in the CLI overwrite those defined in the `.env` file.

bartlomieju commented on cli/args/flags.rs:
```diff
@@ -2563,6 +2571,15 @@ fn import_map_arg() -> Arg {
 
+fn env_file_arg() -> Arg {
+  Arg::new("env-file")
+    .long("env-file")
```
Let's go with `--env`

bartlomieju commented on cli/args/flags.rs:
```diff
@@ -5858,6 +5880,23 @@ mod tests {
 
+  #[test]
+  fn run_env_file() {
+    let r =
+      flags_from_vec(svec!["deno", "run", "--env-file=.env", "script.ts"]);
+    assert_eq!(
```
Please add a test that checks that `deno run --env` will result in `.env` file being the default value.

bartlomieju commented on cli/tools/run.rs:
```diff
@@ -1,5 +1,8 @@
 // Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
 
+extern crate dotenvy;
```
This isn't necessary

bartlomieju commented on cli/tools/run.rs:
```diff
@@ -31,6 +34,10 @@ To grant permissions, set them before the script argument. For example:
     );
   }
 
+  if let Some(env_file_name) = &flags.env_file {
+    from_filename(env_file_name)?;
+  }
```
This should happen in `CliOptions::new()` in `cli/args/mod.rs`

bartlomieju reviewed (commented):
Good start 👍 can you add some integration tests that check that the file is actually loaded?

cknight commented:
How does this change fit in with the dotenv std module?  There is obviously overlap, loading from a .env file.  Some thoughts:

* Is this needed if all this functionality and more can be achieved with a single line of code (import of std/dotenv/load.ts)?
* Is it confusing having this functionality available in both CLI flag and std module?
* Conversely, with a .env.example and .env.defaults files and the ability to access the key/value pairs without exporting to the process environment there is an argument for keeping the std dotenv module as it provides added value which dotenvy doesn't offer
* Do dotenvy and std/dotenv parse a .env file in the same way with the same rules and capabilities?  (Does it matter?)
* I do like the explicit command line reference to loading a .env file,  this makes it more explicit that you are dependent on env variables

bartlomieju commented:
> How does this change fit in with the dotenv std module? There is obviously overlap, loading from a .env file. 
Some thoughts:
> 
> * Is this needed if all this functionality and more can be achieved with a single line of code (import of std/dotenv/load.ts)?

There is some overlap, however with `std/dotenv` you can't influence any of `DENO_*` env vars without spawning a subprocess, while with `--env` flag these env variables will be applied before Deno checks for values of these env vars.

> * Is it confusing having this functionality available in both CLI flag and std module?

Probably yes, however `--env` is way more convenient than having to import module from the standard library, plus see previous comment.

> * Conversely, with a .env.example and .env.defaults files and the ability to access the key/value pairs without exporting to the process environment there is an argument for keeping the std dotenv module as it provides added value which dotenvy doesn't offer

We would certainly keep the `std/dotenv` module for people who need more control or flexibility. The `--env` flag is meant to serve the 80-90% use case.

> * Do dotenvy and std/dotenv parse a .env file in the same way with the same rules and capabilities?  (Does it matter?)

Not sure, and I'm not sure if that matters. We will need to check that. It's not 100% agreed that we'll use `dotenvy` crate - AFAIK it can't do variable substitution which is one of the requirements for the `--env` flag. We already have some infrastructure for dealing with env vars in `deno_task_shell` and we might go with fork/inspired implementation there instead of using a 3rd part crate.

> * I do like the explicit command line reference to loading a .env file,  this makes it more explicit that you are dependent on env variables

That was also one of the requirements for this feature - no magic discovery, not walking up the file system. In more detail the requirements for this feature are as follows:

- require `--env` flag, don't traverse up the CWD if the file is not file
- multiline variables support
- variable substitution
- conflict handling - this is the least specified, I think we just need to decide if the conflicting env var should raise an error or overrwrite (or make it configurable via a flag)

cknight commented:
Thanks @bartlomieju, makes sense.

iuioiua commented:
Changes made. Still, work to do. Regarding the requirements:
* [ ] Don't traverse up the CWD - does this just mean throw when an env file with a path above the CWD is defined?
* [ ] Multiline support - not working with `dotenvy` for some reason
* [ ] Variable substitution - need to test, `dotenvy` should do that automatically
* [ ] Conflict handling - let me see if `dotenvy` does that automatically

TL;DR `dotenvy` should take care of the last 3 requirements automatically, but I need to confirm. Do we want to use `deno_task_shell`? If so, able to provide direction?

nayeemrmn commented:
> We also floated the idea that env vars defined in the dotenv file wouldn't require `--allow-env` permission to access

Couldn't there be secrets there? Like API keys to non-production resources.

bartlomieju commented:
Good point. I guess we can skip that part for now and reevaluate in the future.

bartlomieju commented:
@iuioiua did you have a chance to verify if variable substitution and conflict handling is working in `dotenvy`?

iuioiua commented:
Not yet. I should have a chance within the next couple of days. That cool?

iuioiua commented:
> * [ ]  Conflict handling - let me see if `dotenvy` does that automatically

`dotenvy` doesn't throw on conflicting environment variables. Instead, only the first value is used. I think this behaviour is fine, as long as it's explained, which I included. WDYT?

cknight commented:
I think the more common use case would be if the env variable set in .env was already in the process environment, rather than the use case of duplication in the .env file.  I think the behaviour is the same (the process env value is not overwritten), but it might be worth explicitly stating this in the docs.

dsherret commented on cli/args/flags.rs:
```diff
@@ -391,6 +391,7 @@ pub struct Flags {
   pub ext: Option<String>,
   pub ignore: Vec<PathBuf>,
   pub import_map_path: Option<String>,
+  pub env_file: Option<String>,
```
> That was also one of the requirements for this feature - no magic discovery, not walking up the file system.

Generally a `.env` file would be stored right beside a deno.json. If a deno.json exists, maybe we should traverse up and stop at the deno.json directory when someone provides `--env` with no path?

dsherret commented:
I looked into this a bit and started extracting out some of the code from deno_task_shell, which will help us get this feature exactly how we like it and improve future integration. It's not too complicated, but there are some differences between the variable expansion (see https://github.com/motdotla/dotenv-expand/blob/master/tests/.env). Let's push this off to the next release.

iuioiua commented:
> I think the more common use case would be if the env variable set in .env was already in the process environment, rather than the use case of duplication in the .env file. I think the behaviour is the same (the process env value is not overwritten), but it might be worth explicitly stating this in the docs.

I've updated the command description.

> It's not too complicated, but there are some differences between the variable expansion (see https://github.com/motdotla/dotenv-expand/blob/master/tests/.env).

I've updated tests to match those, which are expectedly now failing. Is that what you'd like us to aim towards, @dsherret?

> I looked into this a bit and started extracting out some of the code from deno_task_shell, which will help us get this feature exactly how we like it and improve future integration.

Can you elaborate on how this is done?

dsherret commented:
>> I looked into this a bit and started extracting out some of the code from deno_task_shell, which will help us get this feature exactly how we like it and improve future integration.
>
> Can you elaborate on how this is done?

Extracted out some of the parsing functions and repurposed them. I haven't worked on it for a few weeks and it currently doesn't compile.

> I've updated tests to match those, which are expectedly now failing. Is that what you'd like us to aim towards, @dsherret?

I think we can just merge this with `dotenvy` as a first pass.

iuioiua commented on cli/args/mod.rs:
```diff
@@ -651,6 +652,15 @@ impl CliOptions {
     let maybe_vendor_folder =
       resolve_vendor_folder(&initial_cwd, &flags, maybe_config_file.as_ref());
 
+    if let Some(env_file_name) = &flags.env_file {
+      from_filename(env_file_name).unwrap_or_else(|_| {
+        panic!(
+          "Unable to load '{}' environment variable file",
+          env_file_name.as_str()
+        )
+      });
+    }
+
```
This block looks wrong to me. How would we like to handle the error?

bartlomieju replied:
Use `bail!` instead which will return an error

iuioiua commented:
This failing test appears to be unrelated to the changes.

iuioiua commented:
Alright, I think this is ready. I've completed up to what I can, for now. I'm unsure how to do the present and missing env file tests for the `compile` and `install` sub-commands or if they're even needed. If they are needed, please go ahead and make any changes necessary. Hopefully, we can land this for the next release.

dsherret reviewed (approved):
LGTM. Thanks @iuioiua!

Let's land this as unstable then circle back with other fixes and improvements in the coming releases.

iuioiua commented:
Nice! Thanks for making the finishing touches. Sorry it took so long 😅
